Add src parameter for uri module that can be used in place of body. Supports binary files

sivel opened the pull request:
##### SUMMARY

This PR adds a new `src` parameter to the `uri` module. The allows for the following behaviors:

* Sending a binary file
* Alternative to `body: "{{ lookup('file', 'foo.json') }}`

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
uri

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
devel
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

sivel committed:
First pass at a src parameter that can be used in place of body. Supports binary files

dminca commented:
I've tested the module and I got the following playbook output:

```sh
TASK [packaging : push OVA archives to Artifactory] *****************************************************************************************************************************************************************************************************************************
fatal: [slave]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'file'. Error was a <type 'exceptions.UnicodeDecodeError'>, original message: 'utf8' codec can't decode byte 0xe0 in position 8205: invalid continuation byte"}

```

The task is pushing 2 `.ova` archives to an Artifactory server, user/pass authentication is initiated.

```yml
- name: push OVA archives to Artifactory                           
  uri:                                                             
    url: "https://example.artifactory.org/{{ item|basename }}"                   
    user: "foobar"                             
    password: "bazbar"                         
    method: PUT                                                    
    body: "{{ lookup('file', playbook_dir +'/artifacts/gateway.ova') }}"                                               
```

sivel commented:
Instead of `body`, you need to use `src`:

```
- name: push OVA archives to Artifactory                           
  uri:                                                             
    url: "https://example.artifactory.org/gateway"                   
    user: "foobar"                             
    password: "bazbar"                         
    method: PUT                                                    
    src: "{{ playbook_dir }}/artifacts/gateway.ova"  
```

dminca commented:
Woopsie, thanks for that one. Fixed now, and retested, the file I uploaded has 642.27 MB

Task:
```yml
- name: push OVA archives to Artifactory                           
  uri:                                                             
    url: "https://example.artifactory.org/virtualbox/{{ item }}"                   
    user: "foobar"                             
    password: "bazbar"                         
    method: PUT                                                    
    src: "{{ playbook_dir }}/artifacts/{{ item }}"
  with_items:
    - gateway.ova
    - slave.ova
```

And the result was

```sh
TASK [packaging : push OVA archives to Artifactory] *****************************************************************************************************************************************************************************************************************************
failed: [slave -> localhost] (item=gateway.ova) => {"changed": false, "content": "", "item": "gateway.ova", "msg": "Status code was not [200]: Connection failure: ('The read operation timed out',)", "redirected": false, "status": -1, "url": "https://example.artifactory.org/virtualbox/gateway.ova"}                                                                 
failed: [gateway -> localhost] (item=slave.ova) => {"changed": false, "content": "", "item": "slave.ova", "msg": "Status code was not [200]: Connection failure: ('The read operation timed out',)", "redirected": false, "status": -1, "url": "https://example.artifactory.org/virtualbox/slave.ova"} 
```

sivel commented:
I'm not convinced there is really any normal timeout happening. You likely need to set a content type for the upload.

```
headers:
    Content-Type: application/gzip
```

I tested a random binary upload to a small webapp I wrote and it succeeded:

```
---
- hosts: localhost
  gather_facts: false
  vars:
    upload_file: /usr/bin/curl
  tasks:
    - stat:
        path: "{{ upload_file }}"
      register: curl_stat

    - uri:
        method: PUT
        url: "http://localhost:5000/{{ upload_file|basename }}"
        src: "{{ upload_file }}"
        return_content: true
        headers:
          Content-Type: application/octet-stream
      register: save

    - stat:
        path: "{{ save.content }}"
      register: upload_stat

    - assert:
        that:
          - curl_stat.stat.checksum == upload_stat.stat.checksum
```

rmauc commented:
Which release can we expect this?
Thanks

sivel commented:
If we can validate that it is working as expected, potentially 2.6.

dminca commented:
Sorry man, still not working on my side, the same error as above reported. I've modified the task, made it like yours, maybe I'm doing something wrong...

sivel committed:
Add test for uri src body

sivel commented:
@dminca I signed up for a trial with artifactory, and was able to get it working with the following adjustments:

```
    - uri:
        url: https://********.jfrog.io/********/generic-local/tinycore-8.2-x86_64.ova
        src: ./ovf/tinycore-8.2-x86_64.ova
        method: PUT
        return_content: true
        user: ********
        password: ********
        force_basic_auth: true
        status_code: 201
        headers:
          Content-Type: application/octet-stream
          Accept: application/json
```

A few arguments to take note from above:

This omission was the ultimate cause to the failure you were seeing:
```
        force_basic_auth: true
```

Status code will be 201:
```
        status_code: 201
```

Python will set a default form content-type, this ensures that doesn't happen, also, the Accept header will make sure we get back a response in JSON that we can read:
```
        headers:
          Content-Type: application/octet-stream
          Accept: application/json
```

I've also just rebased, and added a simple httpbin test.

dminca commented:
Testing now, will let you know the results when done. Thank you for all the effort :+1:

sivel committed:
Bump version_added to 2.6

MauriceS commented:
Used this module to upload a war file to the tomcat manager. Works perfect. Thanks for the PR.

sivel committed:
Close the open file handle

mattclay commented:
@sivel I'm OK with this being included in 2.6.

bcoca commented:
I would make this an action plugin for 'local src' and add a `remote_src` option

sivel commented:
> I would make this an action plugin for 'local src' and add a remote_src option

I suppose that makes sense.  I didn't feel it was necessary.  It does feel a little weird having a `remote_src: True/False` though.  That typically has been left for plugins in the "files" category.  This feels a bit different.

Thoughts?

sivel commented:
@bcoca I don't think I understand your comment. This PR does add src, I was just asking about remote_src because it feels a little out of place to me. I assume you were meaning it should follow the `remote_src: True/False` behavior.

bcoca commented:
my point was that `src` is also prevalent in 'file modules' and this has to do with managing a file to be used by the module, so it makes sense to go all the way and make it behave like `copy: src`, `template: src`, `assemble: src` ... etc.

sivel committed:
Add uri action plugin that handles src/remote_src

* Document remote_src

sivel commented:
@bcoca this now has `remote_src` support via an action plugin. Only took a couple of minutes, as the action plugin for `patch` was nearly identical in need.

bcoca commented on lib/ansible/modules/net_tools/basics/uri.py:
```diff
@@ -132,6 +132,19 @@
+    description:
+      - Path to file to be submitted to the remote server. Cannot be used with I(body).
+        If C(remote_src) is 'no', the module will look for the source file in the
+        I(files) directory and be copied to the target machine before execution.
```
this only happens if a relative path is given and it is not only in 'files' but that would be the first subdir checked.

bcoca commented on lib/ansible/modules/net_tools/basics/uri.py:
```diff
@@ -132,6 +132,19 @@
+      - If C(no), the module will search for src on originating/master machine, if C(yes) the
+        module will use the C(src) path on the remote/target machine.
+    type: bool
+    default: 'no'
```
default of none allows us to know if user set or not, unsure if it is currently needed but JIC we want to make distinction in the future.

sivel replied:
That's just docs, it has no bearing on anything other than letting the docs know the correct value.

sivel committed:
Remove duplicate info about remote_src

sivel committed:
Bump version_added to 2.7

gustavomcarmo commented:
Eager to have it in the next release!

gustavomcarmo commented:
I've just tested and it worked like a charm :)

The playbook makes a backup of all the Docker volumes in a remote host and upload the backup file to a raw Nexus repository. The Nexus instance is behind NGINX.

```yml
---
- hosts: remote
  gather_facts: no
  tasks:
    - name: Make the volumes backup
      archive:
        path: /var/lib/docker/volumes
        dest: volumes.tar.gz
    - name: Make the backup available (requires vault-password-file)
      uri:
        url: https://www.example.com/nexus/repository/backups/volumes.tar.gz
        method: PUT
        src: volumes.tar.gz
        remote_src: true
        user: admin
        password: "{{ nexus_password }}"
        force_basic_auth: yes
        headers:
          Content-Type: application/octet-stream
      vars:
        nexus_password: !vault |
          $ANSIBLE_VAULT;1.1;AES256
          00000000000000000000000000000000000000000000000000000000000000000000000000000000
          00000000000000000000000000000000000000000000000000000000000000000000000000000000
          00000000000000000000000000000000000000000000000000000000000000000000000000000000
          00000000000000000000000000000000000000000000000000000000000000000000000000000000
          0000
      register: upload_result
      failed_when: upload_result.status != 201
      changed_when: upload_result.status == 201
```
